File: back/DevC/src/main/java/com/back/devc/global/security/jwt/JwtAuthenticationFilter.java
package com.back.devc.global.security.jwt;

import com.back.devc.domain.member.member.entity.Member;
import com.back.devc.domain.member.member.entity.MemberStatus;
import com.back.devc.domain.member.member.repository.MemberRepository;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.List;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private static final String AUTHORIZATION_HEADER = "Authorization";
    private static final String BEARER_PREFIX = "Bearer ";

    private final JwtProvider jwtProvider;
    private final MemberRepository memberRepository;

    @Value("${custom.jwt.access-cookie-name:access_token}")
    private String accessCookieName;

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain
    ) throws ServletException, IOException {
        String token = resolveToken(request);

        if (token != null
                && SecurityContextHolder.getContext().getAuthentication() == null
                && jwtProvider.validateAccessTokenStatus(token) == TokenValidationStatus.VALID) {

            Long userId = jwtProvider.getUserId(token);
            Member member = memberRepository.findById(userId).orElse(null);

            if (!isAuthenticatableMember(member)) {
                SecurityContextHolder.clearContext();
                filterChain.doFilter(request, response);
                return;
            }

            String email = member.getEmail();
            String role = member.getRole().name();

            JwtPrincipal principal = new JwtPrincipal(userId, email, role);
            UsernamePasswordAuthenticationToken authentication =
                    new UsernamePasswordAuthenticationToken(
                            principal,
                            null,
                            List.of(new SimpleGrantedAuthority("ROLE_" + role))
                    );

            SecurityContextHolder.getContext().setAuthentication(authentication);
        }

        filterChain.doFilter(request, response);
    }

    // 회원이 존재하고, 상태가 ACTIVE, WARNED, SUSPENDED 중 하나인지 확인
    // 기존: ACTIVE만 가능에서 기능 체크를 위해 수정하였습니다.
    private boolean isAuthenticatableMember(Member member) {
        return member != null &&
                (member.getStatus() == MemberStatus.ACTIVE
                        || member.getStatus() == MemberStatus.WARNED || member.getStatus() == MemberStatus.SUSPENDED);
    }

    private String resolveToken(HttpServletRequest request) {
        String bearerToken = resolveBearerToken(request);
        if (isValidAccessToken(bearerToken)) {
            return bearerToken;
        }

        String cookieToken = resolveCookieToken(request);
        if (isValidAccessToken(cookieToken)) {
            return cookieToken;
        }

        return null;
    }

    private boolean isValidAccessToken(String token) {
        return token != null
                && jwtProvider.validateAccessTokenStatus(token) == TokenValidationStatus.VALID;
    }

    private String resolveBearerToken(HttpServletRequest request) {
        String authorization = request.getHeader(AUTHORIZATION_HEADER);
        if (authorization == null || !authorization.startsWith(BEARER_PREFIX)) {
            return null;
        }

        String token = authorization.substring(BEARER_PREFIX.length()).trim();
        return token.isBlank() ? null : token;
    }

    private String resolveCookieToken(HttpServletRequest request) {
        Cookie[] cookies = request.getCookies();
        if (cookies == null || cookies.length == 0) {
            return null;
        }

        for (Cookie cookie : cookies) {
            if (cookie == null) {
                continue;
            }

            if (accessCookieName.equals(cookie.getName())) {
                String value = cookie.getValue();
                return (value == null || value.isBlank()) ? null : value.trim();
            }
        }

        return null;
    }
}

File: back/DevC/src/main/kotlin/com/back/devc/global/security/jwt/JwtAuthenticationFilter.kt
package com.back.devc.global.security.jwt

import com.back.devc.domain.member.member.entity.Member
import com.back.devc.domain.member.member.entity.MemberStatus
import com.back.devc.domain.member.member.repository.MemberRepository
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.cors.CorsUtils
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException

@Component
class JwtAuthenticationFilter(
    private val jwtProvider: JwtProvider,
    private val memberRepository: MemberRepository,

    @Value("\${custom.jwt.access-cookie-name:access_token}")
    private val accessCookieName: String
) : OncePerRequestFilter() {

    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        return CorsUtils.isPreFlightRequest(request)
    }

    // 모든 요청에서 JWT를 확인하고, 유효한 토큰이면 SecurityContext에 인증 정보를 저장
    // 유효하지 않은 토큰이면 실패 상태를 request attribute에 담아 EntryPoint가 응답 코드로 변환할 수 있게 한다.
    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val tokenResolveResult = resolveToken(request)

        request.setAttribute(
            TOKEN_VALIDATION_STATUS_ATTRIBUTE,
            tokenResolveResult.status
        )

        if (!tokenResolveResult.status.isValid()) {
            SecurityContextHolder.clearContext()
            filterChain.doFilter(request, response)
            return
        }

        if (SecurityContextHolder.getContext().authentication != null) {
            filterChain.doFilter(request, response)
            return
        }

        val token = tokenResolveResult.token
        if (token == null) {
            request.setAttribute(
                TOKEN_VALIDATION_STATUS_ATTRIBUTE,
                TokenValidationStatus.MISSING
            )
            SecurityContextHolder.clearContext()
            filterChain.doFilter(request, response)
            return
        }

        val userId = jwtProvider.getUserId(token)
        val member = memberRepository.findById(userId).orElse(null)

        if (!isAuthenticatableMember(member)) {
            request.setAttribute(
                TOKEN_VALIDATION_STATUS_ATTRIBUTE,
                TokenValidationStatus.INVALID_TOKEN_TYPE
            )
            SecurityContextHolder.clearContext()
            filterChain.doFilter(request, response)
            return
        }

        val memberId = member.userId
        if (memberId == null) {
            request.setAttribute(
                TOKEN_VALIDATION_STATUS_ATTRIBUTE,
                TokenValidationStatus.INVALID_TOKEN_TYPE
            )
            SecurityContextHolder.clearContext()
            filterChain.doFilter(request, response)
            return
        }

        val principal = JwtPrincipal(
            memberId,
            member.email,
            member.role.name
        )

        val authentication = UsernamePasswordAuthenticationToken(
            principal,
            null,
            listOf(SimpleGrantedAuthority("ROLE_${member.role.name}"))
        )

        SecurityContextHolder.getContext().authentication = authentication
        filterChain.doFilter(request, response)
    }

    // 인증 가능한 회원 상태인지 확인
    // 탈퇴 또는 블랙리스트 계정은 유효한 토큰을 가지고 있어도 인증 처리하지 않는다.
    private fun isAuthenticatableMember(member: Member?): Boolean {
        return member != null &&
                (
                        member.status == MemberStatus.ACTIVE ||
                                member.status == MemberStatus.WARNED ||
                                member.status == MemberStatus.SUSPENDED
                        )
    }

    // 요청에서 토큰을 추출하고 검증 결과를 함께 반환
    // Authorization Bearer 토큰을 우선 사용하고, 없으면 access_token 쿠키를 확인
    private fun resolveToken(request: HttpServletRequest): TokenResolveResult {
        val bearerToken = resolveBearerToken(request)

        if (bearerToken != null) {
            val bearerStatus = jwtProvider.validateAccessTokenStatus(bearerToken)
            return TokenResolveResult(bearerToken, bearerStatus)
        }

        val cookieToken = resolveCookieToken(request)

        if (cookieToken != null) {
            val cookieStatus = jwtProvider.validateAccessTokenStatus(cookieToken)
            return TokenResolveResult(cookieToken, cookieStatus)
        }

        return TokenResolveResult(null, TokenValidationStatus.MISSING)
    }

    // Authorization 헤더에서 Bearer 토큰을 추출
    // Bearer 접두어가 없더라도 값이 있으면 잘못된 토큰으로 검증되도록 그대로 반환
    private fun resolveBearerToken(request: HttpServletRequest): String? {
        val authorization = request.getHeader(AUTHORIZATION_HEADER)

        if (authorization.isNullOrBlank()) {
            return null
        }

        if (!authorization.startsWith(BEARER_PREFIX)) {
            return authorization.trim()
        }

        val token = authorization.substring(BEARER_PREFIX.length).trim()

        return token.ifBlank { null }
    }

    // 요청 쿠키에서 access_token 값을 추출
    // Authorization 헤더가 없을 때 쿠키 기반 인증을 처리하기 위해 사용
    private fun resolveCookieToken(request: HttpServletRequest): String? {
        val cookies = request.cookies

        if (cookies.isNullOrEmpty()) {
            return null
        }

        for (cookie in cookies) {
            if (accessCookieName == cookie.name) {
                val value = cookie.value
                return value?.trim()?.ifBlank { null }
            }
        }

        return null
    }

    // 추출한 토큰과 검증 상태를 함께 전달하기 위한 내부 DTO
    // 필터 흐름에서 토큰 문자열과 실패 원인을 분리하지 않고 함께 다루기 위해 사용
    private data class TokenResolveResult(
        val token: String?,
        val status: TokenValidationStatus
    )

    companion object {
        const val TOKEN_VALIDATION_STATUS_ATTRIBUTE = "tokenValidationStatus"
        private const val AUTHORIZATION_HEADER = "Authorization"
        private const val BEARER_PREFIX = "Bearer "
    }
}
